pkg/skaffold/kubernetes/port_forward.go

```diff
@@ -0,0 +1,228 @@
+	resourceVersion int
+	podName         string
+	containerName   string
+	port            int32
+
+	cmd *exec.Cmd
+}
+
+// Forwarder is an interface that can modify and manage port-forward processes
+type Forwarder interface {
+	Forward(*portForwardEntry) error
+	Stop(*portForwardEntry) error
+}
+
+type kubectlForwader struct{}
+
+// Forward port-forwards a pod using kubectl port-forward
+// It returns an error only if the process fails or was terminated by a signal other than SIGTERM
+func (*kubectlForwader) Forward(pfe *portForwardEntry) error {
+	logrus.Debugf("Port forwarding %s", pfe)
+	portNumber := fmt.Sprintf("%d", pfe.port)
+	cmd := exec.Command("kubectl", "port-forward", fmt.Sprintf("pod/%s", pfe.podName), portNumber, portNumber)
+	pfe.cmd = cmd
+
+	buf := &bytes.Buffer{}
+	cmd.Stdout = buf
+	cmd.Stderr = buf
+
+	if err := cmd.Run(); err != nil && !IsTerminatedError(err) {
+		return errors.Wrapf(err, "port forwarding pod: %s, port: %s, err: %s", pfe.podName, portNumber, buf.String())
+	}
+	return nil
+}
+
+// Stop terminates an existing kubectl port-forward command using SIGTERM
+func (*kubectlForwader) Stop(p *portForwardEntry) error {
+	logrus.Debugf("Terminating port-forward %s", p)
+	if p.cmd == nil {
+		return fmt.Errorf("No port-forward command found for %s", p)
+	}
+	if err := p.cmd.Process.Signal(syscall.SIGTERM); err != nil {
+		return errors.Wrap(err, "terminating port-forward process")
+	}
+	return nil
+}
+
+// NewPortForwarder returns a struct that tracks and port-forwards pods as they are created and modified
+func NewPortForwarder(out io.Writer, podSelector PodSelector) *PortForwarder {
+	return &PortForwarder{
+		Forwarder:      &kubectlForwader{},
+		output:         out,
+		podSelector:    podSelector,
+		forwardedPods:  &sync.Map{},
+		forwardedPorts: &sync.Map{},
+	}
+}
+
+func (p *PortForwarder) cleanupPorts() {
+	p.forwardedPods.Range(func(k, v interface{}) bool {
+		entry := v.(*portForwardEntry)
+		if err := p.Stop(entry); err != nil {
+			logrus.Warnf("cleaning up port forwards: %s", err)
+		}
+		return false
+	})
+}
+
+// Start begins a pod watcher that port forwards any pods involving containers with exposed ports.
+// TODO(r2d4): merge this event loop with pod watcher from log writer
+func (p *PortForwarder) Start(ctx context.Context) error {
+	watcher, err := PodWatcher()
+	if err != nil {
+		return errors.Wrap(err, "initializing pod watcher")
+	}
+
+	go func() {
+		defer watcher.Stop()
+
+		for {
+			select {
+			case <-ctx.Done():
+				p.cleanupPorts()
+				return
+			case evt, ok := <-watcher.ResultChan():
+				if !ok {
+					return
+				}
+
+				// Pods will never be "added" in a state that they are ready for port-forwarding
+				// so only watch "modified" events
+				if evt.Type != watch.Modified {
+					continue
+				}
+
+				pod, ok := evt.Object.(*v1.Pod)
+				if !ok {
+					continue
+				}
+				if p.podSelector.Select(pod) && pod.Status.Phase == v1.PodRunning && pod.DeletionTimestamp == nil {
```

Comment:
can we extract the event handling logic and have it covered by unit test? I feel like it's an important piece to protect especially if we want to merge it later with the logging podwatcher.